### PR TITLE
core(fr): prepare emulation utilities for shared use

### PR DESCRIPTION
**Pull request by patrickhulce**
**Summary**
Broken out of #12366, just the emulation shared pieces. We already basically had this split in `lib/emulation.js` so it was just moving the last two methods that put it together :)

**Related Issues/PRs**
ref #11313 #12366

### CODE REVIEW

**Review by connorjclark — APPROVED** [review]

**Review by brendankenny — COMMENTED** [review]
Worth bikeshedding for a minute? Maybe not, considering the status quo :)

But:
```js
emulate()

throttle() // Network and CPU
clearThrottling() // Network and CPU

enableNetworkThrottling()
clearNetworkThrottling()

enableCPUThrottling()
clearCPUThrottling()
```

`enableThrottling()` to match `clearThrottling`/`enableNetworkThrottling`/`enableCPUThrottling`? Or were you thinking in terms of matching `emulate()`?

**Comment by patrickhulce** [comment]
enableThrottling won the informal poll, so we'll rename `throttle` to that

File: lighthouse-core/gather/gather-runner.js [hunk 1]
```diff
@@ -129,7 +130,8 @@ class GatherRunner {
     log.time(status);
     const resetStorage = !options.settings.disableStorageReset;
     await driver.assertNoSameOriginServiceWorkerClients(options.requestedUrl);
-    await driver.beginEmulation(options.settings);
+    await emulation.emulate(driver.defaultSession, options.settings);
+    await emulation.throttle(driver.defaultSession, options.settings);
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/gather/gather-runner.js:134]: Sure, SG

File: lighthouse-core/lib/emulation.js [hunk 2]
```diff
@@ -19,37 +17,61 @@ const NO_CPU_THROTTLE_METRICS = {
...
+ * Sets the throttling options specified in config settings, clearing existing network throttling if
+ * throttlingMethod is not `devtools` (but not CPU throttling, suspected requirement of WPT-compat).
+ *
+ * @param {LH.Gatherer.FRProtocolSession} session
+ * @param {LH.Config.Settings} settings
+ * @return {Promise<void>}
+ */
+async function throttle(session, settings) {
+  // TODO(FR-COMPAT): reconsider if this should be resetting anything
+  if (settings.throttlingMethod !== 'devtools') return clearNetworkThrottling(session);
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/emulation.js:53]: My response here is tightly related to my reflection on the poll so I'll respond to both here.

> enableThrottling won the informal poll, so we'll rename throttle to that

Looking at this more closely, I don't think `enableThrottling` is right. The `emulate` and `throttle` methods aren't blindly turning on the emulation/throttling protocol methods, they're looking at config settings and setting what needs to be set according to the config settings.

A new method called `enableThrottling` that accepts the `throttling` object from settings that just calls the two makes sense, but at that point honestly why bother having a method at all?

I think your argument @brendankenny is that this logic should move up a layer into gather-runner but one of the goals here explicitly was to extract the shared parts for FR and at least for now all of this should be shared logic. Given that, I'm not sure I agree there's a benefit to having this live one layer up (it's not like we're mixing in the `passConfig` object like we were previously, there I wholeheartedly agree it shouldn't live in shared `lib/`).

[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-core/lib/emulation.js:53]: > Looking at this more closely, I don't think `enableThrottling` is right.

conveniently realized _after_ the vote is closed :P

> A new method called `enableThrottling` that accepts the `throttling` object from settings that just calls the two makes sense, but at that point honestly why bother having a method at all?

Ha, you wouldn't have to twist my arm on this if this is the direction you actually wanted to go ;) The `clear*Throttling` methods are basically already there.

I think it's the fact that it's exactly the same condition as `if (passConfig.useThrottling) emulation.throttle(); else emulation.clearThrottling();` that ran right before this line that trips me up. If I were writing this as new code but wanted to match current behavior, I would go with `if (passConfig.useThrottling && settings.throttlingMethod === 'devtools') ...` in the caller in this case (which matches the pass-centric view "in this pass, enable throttling if `throttlingMethod` is `'devtools'` and it's enabled for this pass").

I can easily see the arguments you're making here too, though, so this seems good to go. From the `TODO(FR-COMPAT)`  and `WPT-compat` comments, those conditions might not end up doing the same thing anyways.

**Review by brendankenny — APPROVED** [review]
LGTM

**Comment by erickadalberto** [comment]
> LGTM

**Comment by erickadalberto** [comment]
https://github.com/GoogleChrome/lighthouse/tree/8efad1d24d0a4ee9cf8d881ad7d3727ea7bbbccc

File: lighthouse-core/lib/emulation.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/emulation.js:53]: > I would go with if (passConfig.useThrottling && settings.throttlingMethod === 'devtools')

I'm fine with that. Deal.

[the remark this answers is not in the document]
Comment (brendankenny) [reply lighthouse-core/lib/emulation.js:53]: > It's super specific that we only clear the network but not the CPU when we're doing other throttlingMethods. If that changes in FR I'm happy to move toward your suggestion, but for now that super specific detail feels like it should stay with throttle

oh right, lord. We need to clear up what's going on with that :)

**Pull request merged**